### DISCUSSION TITLE
The plugin isn't sorting the classes when I hit save but it works when npx prettier --write index.html

### DISCUSSION THREAD

**Discussion by codedrian** [tailwindlabs/tailwindcss#10717] [Help] [answered]
**What version of prettier-plugin-tailwindcss are you using?**

0.2.4

**What version of Tailwind CSS are you using?**

3.2.7

**What version of Node.js are you using?**

v18.13.0

**What package manager are you using?**

NPM

**What operating system are you using?**

Ubuntu

**Describe your issue**
I installed the prettier-plugin-tailwindcss and it successfully sorts the Tailwind classes when I run Prettier using npx prettier --write index.html. However, when I hit save in VScode, Prettier works but the prettier-plugin-tailwindcss doesn't sort the Tailwind classes as expected.

I just realized that I didn't set the "prettier.prettierPath" option in my VS Code settings to the path of the Prettier binary installed in my project's node_modules folder.
![image](https://user-images.githubusercontent.com/67692336/222737749-c3055f20-b0c5-458e-880f-3b1de0b7e663.png)

I tried this path but still the plugin is not working as expected when i hit save.
This is my [repository:](https://github.com/Codedrian-Code-Studio/tailwind-craft-station.git)

**Comment by vinaykulk621** [comment]
you need to configure your VS code to format document using prettier
>You can do it by right clicking on the code that you want to format using prettier. you will get a `format document with`  option
![image](https://user-images.githubusercontent.com/94120295/222737241-20baf935-f274-4d3e-89f9-f11c30aeba35.png)

>You can set prettier as default.
![image](https://user-images.githubusercontent.com/94120295/222737339-9f84ef48-77ad-4fec-ad1a-011a81878910.png)

for sort to work execute this command in teminal of your current working directory
```shell
echo {} > .prettierrc.json
```
or create a `.prettierrc.json`  manually and add your prettier configuration init
for example your `.prettierrc.json`
```json
{
  "plugins": ["prettier-plugin-tailwindcss"],
  "tabWidth": 2,
  "trailingComma": "es5",
  "semi": false
}
```

As for `prettier.prettierPath` it is needed if you have installed prettier globaly
above solution is for saving prettier as dev dependency.

**Reply by codedrian** [reply]
I tried that but the plugin didn't work as expected. I attached the video below, I'm stuck in this issue for 1 hour figuring why the plugin didn't work haha.

https://user-images.githubusercontent.com/67692336/222745674-e23c3b84-85d3-4975-a725-c4a7e43ba33d.mp4

**Reply by vinaykulk621** [reply]
So did it work?? it looks like it did work for that div element at the end.

**Reply by codedrian** [reply]
I'm sorry I forgot to mention, the prettier works but the plugin didn't. As you can see the `container` class should go first before the `bg` class. In this vide below I run the prettier using `npx` and the plugin works as expected but when i hit save just like in the fist video above the plugin didn't work.


https://user-images.githubusercontent.com/67692336/222763165-f0a7c526-bb7f-4271-b619-1d8f2765a88d.mp4

**Reply by vinaykulk621** [reply]
My `package.json` file looks like this
```js
  "devDependencies": {
    "autoprefixer": "^10.4.13",
    "dat.gui": "^0.7.9",
    "postcss": "^8.4.21",
    "prettier": "^2.8.4",
    "prettier-plugin-tailwindcss": "^0.2.3",
    "tailwindcss": "^3.2.4",
    "three": "^0.149.0"
  }
```

I ran this command to install that plugin as dev dependency

```shell
npm install -D prettier prettier-plugin-tailwindcss
```
Make sure to update your tailwind version.
it might be because of older version of tailwind.

i see you are missing postcss and autoprefixer.
I suggest you reinstall the tailwindcss again by deleting all the files and packages from your `package.json` file.

**Reply by ch33kaboo** [reply]
Hello there,

Regarding your issue with the prettier-plugin-tailwindcss not sorting the Tailwind classes when you hit save in VS Code, I encountered the same issue recently, and it seems like the problem could be related to the VS Code configuration and to using the latest versions of Prettier and prettier-plugin-tailwindcss.

I got it solved and here is what I did:
- Remove the directory where the VS Code configuration is stored. Note that this will delete any configuration you made on your VS Code or any installed extensions. You can find the directory somewhere in `~/.config`.
- Add `prettier` and `prettier-plugin-tailwindcss` packages to your package.json. However, instead of using the latest versions of the packages, you I tried using older ones that worked for me. For example, the specific versions I used were `"prettier": "^2.7.1"` and `"prettier-plugin-tailwindcss": "^0.1.13"`. To avoid installing the latest versions, you could run `npm install prettier@2.7.1 prettier-plugin-tailwindcss@0.1.13`.
- Make sure you have the Prettier extension installed in VS Code and set it as the default formatter in your settings. Also, make sure the "format on save" option is enabled.

I hope these suggestions help you resolve the issue.

**Reply by gonzaleza61** [reply]
Sorry for the necro thanks for the fix man it worked!

**Reply by Slegendz** [reply]
Thanks @ch33kaboo it worked. Thanks a lot😊🙏

**Comment by karlosbenito** [comment]
use other version and it's work for me with npm install prettier@2.7.1 prettier-plugin-tailwindcss@0.1.13 and restard VS Code

**Comment by r00tb40** [comment]
dont know what the heck with my tailwind sorting plugin, after trying to install older version, remove all the prettier configuration file, remove all plugin in my tailwind.config.js file, it actually works . I use prettier 2.7.1 and tailwind prettier 0.1.13

**Comment by r00tb40** [comment]
I found the solution. first of all you remove all your theme config, and then add 'prettier-plugin-tailwindcss' to plugins, restart vscode and add the theme config again :)

**Reply by Soliha17** [reply]
Yeah, i just restart vscode, working, cheers!

**Reply by wjthompso** [reply]
I'm using VSCode on a Mac and none of the proposed solutions above worked for me. I can get the Tailwind classes to sort when I use `npx prettier --write <path-to-file>` but otherwise it won't sort the classes for me.

**Reply by wjthompso** [reply]
Update: I had to ensure that prettier was the default formatter in VSCode, so I ended up having comment out (the commented out line) and add in this line in my settings.json:

```
    "[typescriptreact]": {
        // "editor.defaultFormatter": "vscode.typescript-language-features"
        "editor.defaultFormatter": "esbenp.prettier-vscode",
        "editor.formatOnSave": true,
        "editor.formatOnType": false,
        "editor.formatOnPaste": false
    },

**Reply by chochodev** [reply]
> I found the solution. first of all you remove all your theme config, and then add 'prettier-plugin-tailwindcss' to plugins, restart vscode and add the theme config again :)

Worked like a charm, thanks a lot

**Comment by ahmafi** [comment]
In case you are using other plugins as well, try changing the order of them, for example:

This doesn't work:
```json
{
  "plugins": ["prettier-plugin-tailwindcss", "prettier-plugin-organize-imports"]
}
```

This works:
```json
{
  "plugins": ["prettier-plugin-organize-imports", "prettier-plugin-tailwindcss"]
}
```

**Reply by Naes0** [reply]
Yep this worked for me too!

```
{
  "plugins": ["prettier-plugin-classnames", "prettier-plugin-tailwindcss"]
}
```

But now the first plugin for classname wrapping doesn't apply 😢

**Reply by Krulknul** [reply]
Just ran into this and was about to give up until I tried this. What the hell

Cheers

**Reply by smac89** [reply]
@Krulknul the kinds of bugs we run into as devs...smh

**Comment by matthiasweiss** [comment]
> In case you are using other plugins as well, try changing the order of them, for example:
> 
> This doesn't work:
> 
> ```json
> {
>   "plugins": ["prettier-plugin-tailwindcss", "prettier-plugin-organize-imports"]
> }
> ```
> 
> This works:
> 
> ```json
> {
>   "plugins": ["prettier-plugin-organize-imports", "prettier-plugin-tailwindcss"]
> }
> ```

Wow, and I've been wondering what I broke...
Thank you!

**Comment by KushajveerSingh** [comment]
I had the same issue, with classes not being sorted. But `npx prettier --write index.html` seemed to be working.

The solution was to add Prettier as the default formatter in VSCode, specifically for HTML files. 
```json
"editor.defaultFormatter": "esbenp.prettier-vscode",
"[html]": {
  "editor.defaultFormatter": "esbenp.prettier-vscode"
},
```

I have no explanation as to why this works, but I wasted an hour trying to figure this out.

Hope this helps.

**Reply by AlexNeep** [reply]
I the same issue where `npx prettier --write .` was working but nothing was happening on save.

My issue was that I had a `.vscode/settings.json` file in my project that was overriding my VSCode settings. 

Updating that (or removing it) to include `"editor.defaultFormatter": "esbenp.prettier-vscode",`, and then restarting worked for me.
